Add time dependences for domain creators

nilsdeppe opened the pull request:
## Proposed changes

Adds what I'm calling `TimeDependence`s for the domain creators.

- `TimeDependence` base class so the specific time dependence can be specified at runtime
- Individual domain creators choose the type of time dependence they want. E.g. a BBH domain probably will have very few to no options for what time dependences are used to generate the moving maps.
- Time dependences are used to retrieve the new maps for individual blocks. Later I will add `inject` functions to the `Block` class that allows adding the time dependence there.
- Add None, UniformTranslation, and Composition time dependences.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add TimeDependence and UniformTranslation

nilsdeppe committed:
Add Composition TimeDependence

nilsdeppe committed:
Add None TimeDependence to signal mesh isn't moving

nilsdeppe committed:
Add is_none to TimeDependence

Allows checking if the TimeDepedence is None, since that signals there is no
time dependence.

moxcodes commented on src/Domain/Creators/TimeDependence/UniformTranslation.hpp:
```diff
@@ -0,0 +1,140 @@
+namespace domain {
+namespace creators {
+namespace time_dependence {
+/// \brief A uniform translation in the \f$x-, y-\f$ and \f$z-\f$direction.
```
I think the documentation for the class should clarify that what is meant by 'uniform' is that it is constant across spatial points, but linearly increasing in time (i.e. a constant velocity). Perhaps consider renaming the class something with `Velocity`?

kidder replied:
hmm I'm pretty sure that this motion is called a uniform translation...  Additional documentation to make it clear is fine, but I don't like renaming the class

moxcodes commented on src/Domain/Creators/TimeDependence/UniformTranslation.hpp:
```diff
@@ -0,0 +1,140 @@
+
+  /// Returns the map for each block to be used in a composition of
+  /// `TimeDependence`s.
+  MapForComposition map_for_composition() const noexcept;
```
why isn't this one in the abstract base interface? will not all `TimeDependence`s be composable like this?

nilsdeppe replied:
Because we don't know the derived class inside the base class and so don't know what to return. 

No, not all will be generically composable. I expect most to actually not be, but that rather we implement specific optimized cases for specific domains we use a lot. The generic ones are just going to be things like a rotation and translation. For anything like the binaries we want to make sure all the maps are optimized, and so wouldn't be a composition of a bunch of different time dependences.

moxcodes commented on tests/Unit/Domain/Creators/TimeDependence/Test_UniformTranslation.cpp:
```diff
@@ -0,0 +1,349 @@
+  CAPTURE(initial_time);
+  CAPTURE(f_of_t_names);
+
+  const auto* const time_dep = dynamic_cast<const UniformTranslation<MeshDim>*>(
```
why do you need to get the `time_dep` out of the unique pointer? Does the `get` not give you something directly that you can compare to `nullptr`?
(also done below on line 98)

nilsdeppe replied:
Ah, this is to test that the factory creation spat out the correct class. To maximally reuse code I decided to stick that check in here. Does that make it clear? I can add a comment if you like :)

moxcodes replied:
ohhh, yeah, please add a code comment for that.

moxcodes commented on tests/Unit/Domain/Creators/TimeDependence/Test_UniformTranslation.cpp:
```diff
@@ -0,0 +1,349 @@
+  // For a random point at a random time check that the values agree. This is to
+  // check that the internals were assigned the correct function of times.
+  UniformCustomDistribution<double> dist{-1.0, 1.0};
+  const auto grid_coords_dv =
```
how about `vector` rather than `dv`

nilsdeppe replied:
I'm indifferent, but we have been using `dv` to represent `DataVector` quite a bit in our tests, which is why I chose this suffix. Let me know what you think :)

moxcodes replied:
I'd prefer `vector`, but won't insist on it given that it's a pattern elsewhere.

moxcodes commented on tests/Unit/Domain/Creators/TimeDependence/Test_UniformTranslation.cpp:
```diff
@@ -0,0 +1,349 @@
+          make_not_null(&gen), make_not_null(&dist), DataVector{5});
+  const auto grid_coords_double =
+      make_with_random_values<tnsr::I<double, MeshDim, Frame::Grid>>(
+          make_not_null(&gen), make_not_null(&dist), 5.0);
```
I think you no longer need to pass the bogus `double` for this to work. (same two lines below)

moxcodes commented on tests/Unit/Domain/Creators/TimeDependence/Test_UniformTranslation.cpp:
```diff
@@ -0,0 +1,349 @@
+  const auto inertial_coords_double =
+      make_with_random_values<tnsr::I<double, MeshDim, Frame::Inertial>>(
+          make_not_null(&gen), make_not_null(&dist), 5.0);
+  for (const auto& block_map : block_maps) {
```
The coordinate map abstract base class seems to have a virtual function that evaluates equality between maps. Is that not sufficient for testing that the maps are the same?

nilsdeppe replied:
We've been burned by incorrect `operator==` implementations, so I've gotten a bit paranoid... But having said that I will add the call and a comment why we test both ways.

moxcodes commented on src/Domain/Creators/TimeDependence/Composition.hpp:
```diff
@@ -0,0 +1,137 @@
+};
+
+template <typename MapsList>
+using generate_coordinate_map_t =
```
can this detail utility be used to simplify the types in the `UniformTranslation` as well?

moxcodes commented on src/Domain/Creators/TimeDependence/Composition.hpp:
```diff
@@ -0,0 +1,137 @@
+  static_assert(
+      tmpl::all<
+          tmpl::integral_list<size_t, TimeDependenceCompTags::mesh_dim...>,
+          std::is_same<tmpl::integral_constant<size_t, mesh_dim>,
```
`cpp17::is_same_v`?

nilsdeppe replied:
The value is on the result of `tmpl::all` not on the `is_same` same so this can't be changed

moxcodes commented on src/Domain/Creators/TimeDependence/Composition.tpp:
```diff
@@ -0,0 +1,146 @@
+}
+
+template <typename TimeDependenceCompTag0, typename... TimeDependenceCompTags>
+Composition<TimeDependenceCompTag0, TimeDependenceCompTags...>::Composition(
```
It would be easier to read the overload logic if the second constructor immediately followed the first in the file.

moxcodes commented on src/Domain/Creators/TimeDependence/Composition.tpp:
```diff
@@ -0,0 +1,146 @@
+namespace detail {
+template <typename... MapsSoFar>
+auto combine_coord_maps(
+    CoordinateMap<Frame::Grid, Frame::Inertial, MapsSoFar...>
```
It seems like these are copying the maps a lot. They're probably pretty small most of the time, but it would still be better to move or reference wherever possible.

moxcodes commented on src/Domain/Creators/TimeDependence/Composition.tpp:
```diff
@@ -0,0 +1,146 @@
+              *rest_time_dep)
+              .map_for_composition()...)),
+      functions_of_time_(detail::combine_functions_of_time(
+          {TimeDependenceCompTag0::name(), TimeDependenceCompTags::name()...},
```
`db::tag_name`?

moxcodes commented on src/Domain/Creators/TimeDependence/Composition.tpp:
```diff
@@ -0,0 +1,146 @@
+  std::unordered_map<std::string,
+                     std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
+      functions_of_time{};
+  size_t counter_into_dep_names = 0;
```
I can't find where this gets incremented

moxcodes commented on src/Domain/Creators/TimeDependence/Composition.tpp:
```diff
@@ -0,0 +1,146 @@
+                                                           FunctionOfTime>>&&
+                                       functions_of_time_single_map) noexcept {
+    for (auto& name_and_func : functions_of_time_single_map) {
+      if (UNLIKELY(functions_of_time.count(name_and_func.first) != 0)) {
```
Sorry, I'm sure I've asked this before, but can you remind me why we are using a map rather than some form of type-indexed object (like a tagged tuple) for the functions of time storage?

nilsdeppe replied:
Because we don't know what functions of time exist until runtime. The reason for this is that coordinating the control system and the domain is really non-trivial and we couldn't figure out a good way of doing it so the decision was made to first get moving mesh and control systems done and working, then consider revisiting if something comes along at that point that could work. Does that clear things up?

moxcodes replied:
Yeah, that makes sense. Thanks!

moxcodes commented on src/Domain/Creators/TimeDependence/Composition.tpp:
```diff
@@ -0,0 +1,146 @@
+  std::unordered_map<std::string,
+                     std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
+      result{};
+  for (const auto& name_and_function_of_time : functions_of_time_) {
```
seems like the copy of an `unordered_map` of `unique_ptr` might be done frequently. If you anticipate repeating this same procedure more than the couple of times in this file, consider creating a helper function for it.

moxcodes commented on tests/Unit/Domain/Creators/TimeDependence/Test_Composition.cpp:
```diff
@@ -0,0 +1,249 @@
+      Composition<TimeDependenceCompositionTag<UniformTranslation<3>>,
+                  TimeDependenceCompositionTag<UniformTranslation<3>, 1>>;
+
+  const std::array<double, 3> velocity0{{2.4, -7.3}};
```
only two numbers given to the 3d velocity

moxcodes commented on tests/Unit/Domain/Creators/TimeDependence/Test_Composition.cpp:
```diff
@@ -0,0 +1,249 @@
+
+  // For a random point at a random time check that the values agree. This is to
+  // check that the internals were assigned the correct function of times.
+  UniformCustomDistribution<double> dist{-1.0, 1.0};
```
much of this seems pretty similar to the test for the previous commit. Can it be combined into a shared test helper?

moxcodes commented on tests/Unit/Domain/Creators/TimeDependence/Test_Composition.cpp:
```diff
@@ -0,0 +1,249 @@
+          gen, make_not_null(&dist), DataVector{5});
+  const auto grid_coords_double =
+      make_with_random_values<tnsr::I<double, MeshDim, Frame::Grid>>(
+          gen, make_not_null(&dist), 5.0);
```
the unused `5.0` argument doesn't have to be included anymore I think.

moxcodes commented on src/Domain/Creators/TimeDependence/None.cpp:
```diff
@@ -0,0 +1,73 @@
+    domain::CoordinateMapBase<Frame::Grid, Frame::Inertial, MeshDim>>>
+None<MeshDim>::block_maps(const size_t /*number_of_blocks*/) const noexcept {
+  ERROR(
+      "The 'block_maps' function of the 'None' TimeDepnedence should never be "
```
`TimeDepnedence`->`TimeDependence` (and below)

kidder commented on src/Domain/Creators/TimeDependence/TimeDependence.hpp:
```diff
@@ -0,0 +1,81 @@
+///
+/// The simplest examples of a `TimeDependence` are `None` and
+/// `UniformTranslation`. The `None` class is treated in a special manner to
+/// communicate to the code that the domain is time-dependent.
```
time-independent?

kidder commented on src/Domain/Creators/TimeDependence/UniformTranslation.hpp:
```diff
@@ -0,0 +1,140 @@
+  static constexpr size_t mesh_dim = MeshDim;
+
+  /// \brief The initial time of the functions of time.
+  struct InitialTime {
```
this is the time at which the map is the identity?  what makes sure it is specified to the initial time of the evolution?

nilsdeppe replied:
Yes, nothing. We can't do that with the way the option parsing currently works.

kidder commented on src/Domain/Creators/TimeDependence/Composition.hpp:
```diff
@@ -0,0 +1,137 @@
+ * that is a composition of existing `TimeDependences`.
+ *
+ * The first template parameter is the existing TimeDependence while the
+ * `Suffix` parameter can be used by a composition contains multiple of the same
```
composition that contains more than one of the same

kidder replied:
ping

nilsdeppe committed:
Add clone_unique_ptr function

For moving mesh we often need to clone/copy unordered_maps and vector (maybe
other data structures in the future) of unique_ptrs.

nilsdeppe commented:
Thanks for the reviews! Rebased on develop and pushed fixups :)

nilsdeppe commented:
I also added a new commit `Add clone_unique_ptr function` to the beginning of the history (ignore where github thinks it is)

kidder commented on tests/Unit/Domain/Creators/TimeDependence/Test_UniformTranslation.cpp:
```diff
@@ -113,20 +117,12 @@ void test(const std::unique_ptr<TimeDependence<MeshDim>>& time_dep_unique_ptr,
+  TIME_DEPENDENCE_GENERATE_COORDS(make_not_null(&gen), MeshDim, -1.0, 1.0);
+
   for (const auto& block_map : block_maps) {
+    // We've check equivalence above (CHECK(*black_map == expected_block_map);),
```
checked

kidder reviewed (changes requested):
I'm happy for you to squash (including the ping and my other comment) if @moxcodes is

moxcodes commented:
looks great, go ahead and squash.

nilsdeppe commented:
Rebased, squashed, and pushed. Thanks for the reviews!
